### PR TITLE
Add spec for new Interval / IntervalIndex methods: .overlaps(), .covers()

### PR DESCRIPTION
**Pull request by alexlenail**
- [x] part 2/2 for #16316
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Working on parameterizing the tests. Will ping when ready for review.

### CODE REVIEW

File: pandas/tests/indexes/interval/test_interval_new.py [hunk 1]
```diff
@@ -314,3 +314,197 @@ def test_contains_method(self):
 
         assert not index.contains(20)
         assert not index.contains(-20)
+
+    def test_interval_covers_interval(self):
+
+        # class Interval:
+        #     def covers(self, other: Interval) -> bool
+
+        assert Interval(1, 3).covers(Interval(1.5, 2.5))
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval_new.py]: I think `Interval` level tests would be more appropriate in `pandas/tests/scalar/test_interval.py`, which is where the other `Interval` specific tests are located.

File: pandas/tests/indexes/interval/test_interval_new.py [hunk 2]
```diff
@@ -314,3 +314,197 @@ def test_contains_method(self):
...
+        assert Interval(1, 3, closed='left').overlaps(Interval(3, 5, closed='left'))
+        assert Interval(1, 3, closed='left').overlaps(Interval(3, 5, closed='both'))
+
+        # both
+        assert Interval(1, 3, closed='both').overlaps(Interval(1, 3, closed='right'))
+        assert Interval(1, 3, closed='both').overlaps(Interval(1, 3, closed='left'))
+        assert Interval(1, 3, closed='both').overlaps(Interval(1, 3, closed='both'))
+
+        assert Interval(1, 3, closed='both').overlaps(Interval(-1, 1, closed='right'))
+        assert not Interval(1, 3, closed='both').overlaps(Interval(-1, 1, closed='left'))
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval_new.py]: btw I would be very happy to have these in a different file, e.g.

``pandas/tests/indexes/interval/test_overlap_covers.py`` or whatever (excluding @jschendel commet that the Interval tests need to go in the scalar tests); we can make a sub-dir there as well I think.

**Comment by zfrenchee** [comment]
Ready for a look-over on structure @jreback @jschendel. Are the tests in the right place / well structured? Are there missing tests / cases?

File: pandas/tests/scalar/test_interval_new.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+            },
+            'neither': {
+                'right': {(-1, 1): False, (1, 3): True, (3, 5): False},
+                'left':{(-1, 1): False, (1, 3): True, (3, 5): False},
+                'both': {(-1, 1): False, (1, 3): True, (3, 5): False},
+                'neither': {(-1, 1): False, (1, 3): True, (3, 5): False},
+            }
+        }
+
+        assert ivl.overlaps(other) == should_overlap[other_side][ivl_side][ivl_range] == other.overlaps(ivl)
```
Comment (jreback) [inline pandas/tests/scalar/test_interval_new.py]: all of the tests from here on down should go in: ``pandas/tests/indexes/interval/test_overlaps.py`` (new file)

**Comment by jreback** [comment]
what is the reason for *both* covers and overlaps?

**Comment by zfrenchee** [comment]
@jreback (from @shoyer):

> An interval covers another interval if all points in the second interval are found in the first interval.
>
> An interval overlaps another interval if there exist any points found in both intervals.
-- (https://github.com/pandas-dev/pandas/issues/16316#issuecomment-301352130)

I personally am interested in overlaps, but I can imagine people might want to ask what intervals are covered by their IntervalIndex or vice-versa. Does that answer the question?

**Comment by jschendel** [comment]
> what is the reason for both covers and overlaps?

FWIW other implementations of `Interval` have both methods, e.g. the `intervaltree` package on pypi, and `range` types in postgres, albeit with a slightly different naming convention for `covers`.  The `intervaltree` package uses a `contains_interval` method for `covers`.  In postgres it's just an operator, `@>`,  which is described in the docs as "contains range".  I suppose we could maybe rename `covers` to be more in line with these, but I don't really have a strong opinion either way.

Is there any utility in adding a `strict` kwarg to `covers`?  By that I basically mean `strict=False` is existing behavior (default), `strict=True` is existing behavior and no shared endpoints:
```python
In [1]: iv = Interval(0, 1)

In [2]: iv.covers(iv, strict=False)
Out[2]: True

In [3]: iv.covers(iv, strict=True)
Out[3]: False
```
Could probably wait to see if there's demand for `strict` prior to implementing, unless there's a reason anyone wants it immediately implemented.

**Comment by jreback** [comment]
rebase and push again, fixed some hanging by Travis CI

**Comment by gfyoung** [comment]
@zfrenchee : Looks like you have some lint errors.  Otherwise, Travis looks happy to me.

**Comment by zfrenchee** [comment]
I'd like to revisit the interface we agreed upon (initially from @shoyer) which this PR introduces tests for. The interface was

```
class Interval:
    def covers(self, other: Interval) -> bool
    def covers(self, other: IntervalIndex) -> IntegerArray1D
    def overlaps(self, other: Interval) -> bool
    def overlaps(self, other: IntervalIndex) -> IntegerArray1D

class IntervalIndex:
    def covers(self, other: Interval) -> IntegerArray1D
    def covers(self, other: IntervalIndex) -> Tuple[IntegerArray1D, IntegerArray1D]
    def overlaps(self, other: Interval) -> IntegerArray1D
    def overlaps(self, other: IntervalIndex) -> Tuple[IntegerArray1D, IntegerArray1D]
```

This is the last call for modifications to that interface in this PR. @jreback @jorisvandenbossche @gfyoung . What do you think?

**Comment by shoyer** [comment]
Thinking about this again: there's some ambiguity about whether coverage for `IntervalIndex.covers(IntervalIndex)`  is:
- elementwise: does each interval in the first index cover each corresponding interval?
- all: does each interval in the first index cover *all* of the second intervals?
- any: does each interval in the first index cover *any* of the second intervals?

Most pandas operations work elementwise, so arguably that's clearest. That would suggest we should only return booleans/boolean-arrays, i.e.,
```
class Interval:
    def covers(self, other: Interval) -> bool
    def covers(self, other: IntervalIndex) -> BoolArray1D
    def overlaps(self, other: Interval) -> bool
    def overlaps(self, other: IntervalIndex) -> BoolArray1D

class IntervalIndex:
    def covers(self, other: Interval) -> BoolArray1D
    def covers(self, other: IntervalIndex) -> BoolArray1D
    def overlaps(self, other: Interval) -> BoolArray1D
    def overlaps(self, other: IntervalIndex) -> BoolArray1D
```

If we adopted this choice, we would save "covers all", "covers any", "overlaps all" and "overlaps any" for another method, which might be implemented in the future (if necessary).

**Comment by zfrenchee** [comment]
Thanks for those comments @shoyer. 

I'm personally rather attached to your original suggestion, because I think all these other methods can be expressed with short recipes derived from the methods in their current form in the spec. I'm not sure anyone will really be asking the question "which intervals in IntervalIndex A cover _every_ interval in IntervalIndex B?" and even if they did want to ask that, using the current spec they could get indices back and then check if that index set is complete (2 lines instead of 1). The reverse doesn't work -- If we provide these methods, we can't ask the more basic question "which intervals from A cover which intervals from B?" 

Are you thinking of a particular other method in the current pandas API to base your suggestion from? If so we have to weigh consistency as a factor. Let me know what you think.

File: pandas/tests/scalar/test_interval_new.py (hunk 3 above)
Comment (zfrenchee) [reply pandas/tests/scalar/test_interval_new.py]: @jreback do you still want to see this?

**Comment by zfrenchee** [comment]
@pep8speaks

**Comment by zfrenchee** [comment]
@jreback @gfyoung I believe this is ready for merge.

File: pandas/tests/scalar/test_interval_new.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            },
+            'neither': {
+                'right': {(-1, 1): False, (1, 3): True, (3, 5): False},
+                'left':{(-1, 1): False, (1, 3): True, (3, 5): False},
+                'both': {(-1, 1): False, (1, 3): True, (3, 5): False},
+                'neither': {(-1, 1): False, (1, 3): True, (3, 5): False},
+            }
+        }
+
+        assert ivl.overlaps(other) == \
```
Comment (jreback) [inline pandas/tests/scalar/test_interval_new.py]: these are very hard to read and debug much better to do
```
result = ...
expected = ...
assert result == expected
```

**Review by jreback — CHANGES_REQUESTED** [review]
you need to have Interval tests in pandas/tests/scalar/, but IntervalIndex tests in pandas/tests/indexes/interval

**Comment by zfrenchee** [comment]
@jreback 

```
~/Documents/pandas(master) » git remote add upstream https://github.com/pandas-dev/pandas.git
------------------------------------------------------------
~/Documents/pandas(master) » git diff upstream/master -u -- "*.py" | flake8 --diff                                                     alex@alexs-mbp
fatal: bad revision 'upstream/master'
```

**Comment by jreback** [comment]
@zfrenchee your upstream may be called something else, use ``git remote -v`` to see

**Comment by jreback** [comment]
I think you have to tell autopep to fix the lines, its not automatic:

``autopep8 --max-line-length``

note that we generally *don't* like to use autopep for line breaking. we do this manually (because autopep8  produces weird output).

Its not that much code you can easily fix it manually.

**Comment by zfrenchee** [comment]
@jreback green.

**Comment by shoyer** [comment]
Before we merge this, can someone else weigh in on the API options here? See https://github.com/pandas-dev/pandas/pull/18975#issuecomment-354787556 and the comment below

**Comment by jorisvandenbossche** [comment]
> Most pandas operations work elementwise, so arguably that's clearest. That would suggest we should only return booleans/boolean-arrays, i.e.,

I agree with this. For me it looks very strange that a Index operation returns a tuple.

But I find it very hard to think about this without code. @zfrenchee could you write up a few hypothetical example of how you would use it? To have a bit more the idea how it would look like.

As I asked for in the other PR as well, I think it would also be helpful to clear out the design to actually write out the reference docs (the docstrings of those methods). We can eg add them now as a multi-line comment at the top of the test files. BTW, I think it would also be rather easy to actually implement the methods. That might make it easier to experiment with.

**Comment by jschendel** [comment]
> can someone else weigh in on the API options here

I agree with @shoyer that returning booleans/boolean-arrays makes more sense for `overlaps` and `covers`.  I plan to implement an interval accessor (#16401) in the near future now that the new accessor interface PR has been merged.  I haven't looked too closely at the implementation, but I _think_ returning booleans would enable indexing syntax along the lines of `df[df.interval_col.iv.overlaps(other)]`.  At the very least it seems like it would make this behavior consistent with the accessor.

That being said, I'm not opposed to adding the functionality suggested by @zfrenchee, but it might be better served with a different name.  Perhaps with a prefix along the lines of `get_*`, e.g. `IntervalIndex.get_overlaps(other)`, or maybe even a more generic method with a `how` parameter, e.g. `IntervalIndex.something_generic(other, how='overlaps|covers|...')`.

Unrelated, but one thing that stands out to me is `Interval.overlaps(IntervalIndex)` (likewise `covers`).  Looks like this is identical to `IntervalIndex.overlaps(Interval)`, so having both seems a little redundant.  Syntactically `Interval.overlaps(IntervalIndex)` feels a little strange to me, since you have a method on a scalar taking a non-scalar argument.  I'd be fine removing `Interval.overlaps(IntervalIndex)` and just implementing `IntervalIndex.overlaps(Interval)` (likewise `covers`), but I don't have particularly strong feelings either way.

**Comment by zfrenchee** [comment]
> But I find it very hard to think about this without code. @zfrenchee could you write up a few hypothetical example of how you would use it? To have a bit more the idea how it would look like.

I'm hoping to do genome arithmetic, like [BEDtools](http://bedtools.readthedocs.io/en/latest/). Check out the [BEDtools tools](http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html), e.g. [intersect](http://bedtools.readthedocs.io/en/latest/content/tools/intersect.html), [merge](http://bedtools.readthedocs.io/en/latest/content/tools/merge.html), and [closest](http://bedtools.readthedocs.io/en/latest/content/tools/closest.html). With the current implementation of overlaps (in order to do something like bedtools intersect) I would do
```
one, two = idx1.overlaps(idx2)
return pd.concat((idx1[one], idx2[two]), axis=1)
```

I think that the issue with the boolean overlaps function is that it's hard to trace what overlaps what. E.g. consider: 

idx1 = IntervalIndex.from_tuples([(1, 5), (2, 6)])
idx2 = IntervalIndex.from_tuples([(0, 4), (3, 7)])

then 
```
pd.concat((idx1[idx1.overlaps(idx2)], idx2[idx2.overlaps(idx2)]))
```
misses half of the overlaps. Do you see what I mean? Let me know if that needs to be clearer. Basically, if it were booleans, you wouldn't necessarily get all pairs of overlaps out from the overlaps function.

> I agree with this. For me it looks very strange that a Index operation returns a tuple.

I can't deny it looks off. However, for this index, for this operation, it might be the only thing that makes sense? Very open to alternatives...

> Unrelated, but one thing that stands out to me is Interval.overlaps(IntervalIndex) (likewise covers). Looks like this is identical to IntervalIndex.overlaps(Interval), so having both seems a little redundant. Syntactically Interval.overlaps(IntervalIndex) feels a little strange to me, since you have a method on a scalar taking a non-scalar argument.

This is a good point. [We discussed this a little.](https://github.com/pandas-dev/pandas/issues/16316#issuecomment-354573878) I also don't have strong feelings, and agree that if we were to remove one of them, `Interval.overlaps(IntervalIndex)` would be the one. I'd love to find some other useful functionality for that function signature, but I wasn't able to come up with one.

@shoyer @jreback @jschendel @jorisvandenbossche

**Comment by zfrenchee** [comment]
bump @jreback @shoyer @jschendel @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
@zfrenchee I am still trying to understand the logic behind `one, two = idx1.overlaps(idx2)` (I mean what it does exactly). Can you show what the output would be for the example `idx1` and `idx2` ?

**Comment by jorisvandenbossche** [comment]
(sidenote: if we decide something like this it too complicated or inconsistent for as a method, we should make sure it is not too hard to write a function that does that)

**Comment by zfrenchee** [comment]
@jorisvandenbossche: @shoyer 's original API suggestion, which I add tests for in this PR:

```
>>> idx1 = IntervalIndex.from_tuples([(1, 5), (2, 6)])
>>> idx2 = IntervalIndex.from_tuples([(0, 4), (3, 7)])
>>> idx1.overlaps(idx2)
([0, 0, 1, 1], [0, 1, 0, 1])
```
as opposed to 
```
>>> idx1 = IntervalIndex.from_tuples([(1, 5), (2, 6)])
>>> idx2 = IntervalIndex.from_tuples([(0, 4), (3, 7)])
>>> idx1.overlaps(idx2)
[true, true]
```

**Comment by zfrenchee** [comment]
I'm alright renaming the method `.find_overlaps` if that reduces the confusion for not returning boolean masks, or even (maybe) setting this as a method on dataframes with IntervalIndex indexes, rather than on the indexes themselves... I just really want this functionality!

**Comment by jschendel** [comment]
> I am still trying to understand the logic behind `one, two = idx1.overlaps(idx2)` (I mean what it does exactly)

@jorisvandenbossche : IIUC the "plain english" interpretation is, looking at the two lists pairwise, specifically which interval in `idx1` overlaps which interval in `idx2`.  For example, in the output that @zfrenchee just posted:
```python
>>> idx1 = IntervalIndex.from_tuples([(1, 5), (2, 6)])
>>> idx2 = IntervalIndex.from_tuples([(0, 4), (3, 7)])
>>> idx1.overlaps(idx2)
([0, 0, 1, 1], [0, 1, 0, 1])
```
The first pairwise element, `(0, 0)`, says that `idx1[0]` overlaps `idx2[0]`.  And likewise for the other pairs, `(0, 1)`, `(1, 0)`, and `(1, 1)`.

**Comment by zfrenchee** [comment]
Thanks @jschendel. Sorry if I wasn't clear, @jorisvandenbossche.

**Comment by jorisvandenbossche** [comment]
OK, so you basically do a cross product combination, and then filter for the overlapping ones? So the length of the output varies depending on the data. 
Personally I think we should not call that `overlaps`, but another method like you propose might be a good idea.

(this discussion is another reason I think it would be useful to actually write good docstrings for the functions your propose or even proof of concept implementation, that will makes things a lot easier to discuss)

**Comment by jorisvandenbossche** [comment]
BTW, you could see this as a 'join' based on the overlap. And you could then also join based on other ops like within or contains (like with spatial joins).

Would it actually be more useful to directly join two frames based on this? Instead of providing the indices with which you can then do the concat manually? (or are there case you rather want the indices and not directly the merged frames?)

**Comment by zfrenchee** [comment]
@jorisvandenbossche 

> Would it actually be more useful to directly join two frames based on this? Instead of providing the indices with which you can then do the concat manually? (or are there case you rather want the indices and not directly the merged frames?)

I think there could be use cases for getting back the indices without doing the join, so I would vote to keep this as a two-step process. You might want to iterate through the zip of the result, and ask questions about each overlaps. You might want to just look at one side or the other. I personally will mostly likely just be using these functions solely for join operations via `concat`, but I think it's quite feasible others would have different uses for this.

**Comment by zfrenchee** [comment]
> Personally I think we should not call that `overlaps`, but another method like you propose might be a good idea.

I'm happy to name it anything, but I can't think of a better name: The problem this method solves is finding the overlaps between two lists of intervals, i.e. two IntervalIndexes. What else would you call it?

**Comment by zfrenchee** [comment]
@jschendel @jorisvandenbossche @jreback

**Comment by jorisvandenbossche** [comment]
See also my comment above https://github.com/pandas-dev/pandas/pull/18975#issuecomment-359202598

File: pandas/tests/indexes/interval/test_interval_overlaps_covers.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+class TestIntervalIndex(Base):
+
+    def _compare_tuple_of_numpy_array(self, result, expected):
+        lidx, ridx = result
+        lidx_expected, ridx_expected = expected
+
+        tm.assert_numpy_array_equal(lidx, lidx_expected)
+        tm.assert_numpy_array_equal(ridx, ridx_expected)
+
+    def test_intervalIndex_covers_intervalIndex(self):
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval_overlaps_covers.py]: can you parameterize

File: pandas/tests/indexes/interval/test_interval_overlaps_covers.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        self._compare_tuple_of_numpy_array(result, expected)
+
+        idx3 = IntervalIndex.from_tuples([(0, 1), (2, 3), (1, 3)],
+                                         closed="both")
+
+        result = idx.covers(idx3),
+        expected = (np.array([0, 1, 2, 2]), np.array([0, 1, 1, 2]))
+        self._compare_tuple_of_numpy_array(result, expected)
+
+    def test_intervalIndex_overlaps_intervalIndex(self):
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval_overlaps_covers.py]: same

File: pandas/tests/scalar/test_interval_new.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        tm.assert_numpy_array_equal(other_result, expected)
+
+    @pytest.mark.parametrize("idx_side", ['right', 'left', 'both', 'neither'])
+    @pytest.mark.parametrize("ivl_side", ['right', 'left', 'both', 'neither'])
+    @pytest.mark.parametrize("ivl_range", [(1, 3), (1, 2), (0, 2), (3, 4)])
+    def test_interval_overlaps_intervalIndex(self, idx_side, ivl_side,
+                                             ivl_range):
+
+        # class Interval:
+        #     def overlaps(self, other: IntervalIndex) -> IntegerArray1D
```
Comment (jreback) [inline pandas/tests/scalar/test_interval_new.py]: the II tests need to go in pandas/tests/indexes/interval/test_interval_new.py

File: pandas/tests/scalar/test_interval_new.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        assert Interval(1, 3).covers(Interval(1.5, 2.5))
+        assert Interval(1, 3).covers(Interval(1, 2))
+        assert Interval(1, 3).covers(Interval(2, 3))
+        assert not Interval(1, 3).covers(Interval(0.5, 2.5))
+        assert not Interval(1, 3).covers(Interval(1.5, 3.5))
+
+        ivl = Interval(1, 3, closed=ivl_side)
+        other = Interval(1, 3, closed=oth_side)
+
+        should_cover = {
```
Comment (jreback) [inline pandas/tests/scalar/test_interval_new.py]: would like to ideally enocde these results IN the parameterization if possible.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
we had a lot of good discussion about this, but is pretty stale now.

**Pull request closed without merging**